File: src/lib/wow/archive/casc/cdn-resolver.ts
/**
 * CDN host resolution with intelligent pre-caching, ported from wow.export
 * (src/js/casc/cdn-resolver.js).
 */
import util from 'util';

import { write } from '@/lib/wow/log';

import { constants } from '../../formats/constants';
import { get, ping } from '../../formats/generics';
import { parseVersionConfig, VersionConfigEntry } from './version-config';

interface RankedHost {
  host: string;
  ping: number;
}

interface ResolutionCacheEntry {
  promise: Promise<RankedHost[]> | null;
  bestHost: RankedHost | null;
  rankedHosts?: RankedHost[] | null;
}

class CDNResolver {
  /** Map of cacheKey -> { promise, bestHost }. cacheKey = region + '|' + hosts. */
  private resolutionCache = new Map<string, ResolutionCacheEntry>();

  /** Track hosts that have failed to respond properly (e.g., censored responses). */
  private failedHosts = new Set<string>();

  /** Start pre-resolution for a region if not already started. */
  startPreResolution(region: string, product = 'wow'): void {
    write('Starting CDN pre-resolution for region: %s', region);
    void this._resolveRegionProduct(region, product);
  }

  /** Get the best host for a region with specific server config. */
  async getBestHost(region: string, serverConfig: VersionConfigEntry): Promise<string> {
    const cacheKey = this._getCacheKey(region, serverConfig.Hosts);
    const cached = this.resolutionCache.get(cacheKey);

    if (cached?.bestHost) {
      write('Using cached CDN host for %s: %s', region, cached.bestHost.host);
      return `${cached.bestHost.host}${serverConfig.Path}/`;
    }

    if (cached?.promise) {
      write('Waiting for CDN resolution for %s', region);
      const result = await cached.promise;
      return `${result[0].host}${serverConfig.Path}/`;
    }

    write('Resolving CDN hosts for %s: %s', region, serverConfig.Hosts);
    const promise = this._resolveHosts(region, serverConfig);

    this.resolutionCache.set(cacheKey, { promise, bestHost: null });

    const rankedHosts = await promise;
    this.resolutionCache.set(cacheKey, {
      promise: null,
      bestHost: rankedHosts[0],
      rankedHosts,
    });

    return `${rankedHosts[0].host}${serverConfig.Path}/`;
  }

  /**
   * Get all available hosts for a region ranked by ping speed.
   * Excludes hosts that have previously failed.
   */
  async getRankedHosts(region: string, serverConfig: VersionConfigEntry): Promise<string[]> {
    const cacheKey = this._getCacheKey(region, serverConfig.Hosts);
    const cached = this.resolutionCache.get(cacheKey);

    if (cached?.rankedHosts) {
      write('Using cached ranked CDN hosts for %s', region);
      return cached.rankedHosts.map((h) => `${h.host}${serverConfig.Path}/`);
    }

    if (cached?.promise) {
      write('Waiting for CDN resolution for %s', region);
      await cached.promise;
      const updated = this.resolutionCache.get(cacheKey)!;
      return updated.rankedHosts!.map((h) => `${h.host}${serverConfig.Path}/`);
    }

    write('Resolving CDN hosts for %s: %s', region, serverConfig.Hosts);
    const promise = this._resolveHosts(region, serverConfig);

    this.resolutionCache.set(cacheKey, { promise, bestHost: null, rankedHosts: null });

    const rankedHosts = await promise;
    this.resolutionCache.set(cacheKey, {
      promise: null,
      bestHost: rankedHosts[0],
      rankedHosts,
    });

    return rankedHosts.map((h) => `${h.host}${serverConfig.Path}/`);
  }

  /** Mark a host as failed (e.g., due to censorship or invalid responses). */
  markHostFailed(host: string): void {
    write('Marking CDN host as failed: %s', host);
    this.failedHosts.add(host);
  }

  private _getCacheKey(region: string, hosts: string): string {
    return `${region}|${hosts}`;
  }

  private async _resolveRegionProduct(region: string, product: string): Promise<void> {
    try {
      const host = util.format(constants.PATCH.HOST, region);
      const url = host + product + constants.PATCH.SERVER_CONFIG;
      const res = await get(url);

      if (!res.ok) throw new Error(util.format('HTTP %d from server config endpoint: %s', res.status, url));

      const serverConfigs = parseVersionConfig(await res.text());
      const serverConfig = serverConfigs.find((e) => e.Name === region);

      if (!serverConfig) throw new Error(`CDN config does not contain entry for region ${region}`);

      // Use getBestHost to resolve and cache
      await this.getBestHost(region, serverConfig);
    } catch (error) {
      write('Failed to pre-resolve CDN hosts for region %s: %s', region, (error as Error).message);
    }
  }

  /** Ping all hosts in server config and rank them by speed. */
  private async _resolveHosts(region: string, serverConfig: VersionConfigEntry): Promise<RankedHost[]> {
    write('Resolving best host for %s: %s', region, serverConfig.Hosts);

    const hosts = serverConfig.Hosts.split(' ').map((e) => `https://${e}/`);
    const validHosts: RankedHost[] = [];
    const hostPings: Promise<void>[] = [];

    for (const host of hosts) {
      if (this.failedHosts.has(host)) {
        write('Skipping previously failed host: %s', host);
        continue;
      }

      hostPings.push(ping(host).then((pingMs) => {
        write('Host %s resolved with %dms ping', host, pingMs);
        validHosts.push({ host, ping: pingMs });
      }).catch((e) => {
        write('Host %s failed to resolve a ping: %s', host, e);
      }));
    }

    await Promise.allSettled(hostPings);

    if (validHosts.length === 0) throw new Error('Unable to resolve any CDN hosts (all failed or blocked).');

    validHosts.sort((a, b) => a.ping - b.ping);

    write('%s resolved as the fastest host with a ping of %dms', validHosts[0].host, validHosts[0].ping);
    return validHosts;
  }
}

export const cdnResolver = new CDNResolver();

export default cdnResolver;

File: golang/internal/wow/archive/casc/cdn_resolver.go
package casc

import (
	"fmt"
	"sort"
	"strings"
	"sync"

	"github.com/pqhuy98/wow-converter/internal/wow/constants"
	"github.com/pqhuy98/wow-converter/internal/wow/formats"
	"github.com/pqhuy98/wow-converter/internal/wow/log"
)

type rankedHost struct {
	host string
	ping int
}

type resolutionCacheEntry struct {
	promise     chan struct{}
	bestHost    *rankedHost
	rankedHosts []rankedHost
	done        bool
	err         error
}

// CDNResolver resolves CDN hosts with ping-based ranking.
type CDNResolver struct {
	resolutionCache map[string]*resolutionCacheEntry
	failedHosts     map[string]struct{}
	mu              sync.Mutex
}

// DefaultCDNResolver is the shared CDN resolver instance.
var DefaultCDNResolver = &CDNResolver{
	resolutionCache: map[string]*resolutionCacheEntry{},
	failedHosts:     map[string]struct{}{},
}

// StartPreResolution starts CDN pre-resolution for a region.
func (r *CDNResolver) StartPreResolution(region string, product string) {
	if product == "" {
		product = "wow"
	}
	log.Write("Starting CDN pre-resolution for region: %s", region)
	go func() {
		host := fmt.Sprintf(constants.Patch.Host, region)
		url := host + product + constants.Patch.ServerConfig
		res, err := formats.Get(url)
		if err != nil || !res.OK {
			if err != nil {
				log.Write("Failed to pre-resolve CDN hosts for region %s: %s", region, err.Error())
			}
			return
		}
		serverConfigs := ParseVersionConfig(string(res.Body))
		var serverConfig VersionConfigEntry
		for _, e := range serverConfigs {
			if e["Name"] == region {
				serverConfig = e
				break
			}
		}
		if serverConfig == nil {
			log.Write("Failed to pre-resolve CDN hosts for region %s: CDN config does not contain entry for region %s", region, region)
			return
		}
		_, _ = r.GetBestHost(region, serverConfig)
	}()
}

// GetBestHost returns the best CDN host for a region.
func (r *CDNResolver) GetBestHost(region string, serverConfig VersionConfigEntry) (string, error) {
	cacheKey := r.cacheKey(region, serverConfig["Hosts"])
	r.mu.Lock()
	cached := r.resolutionCache[cacheKey]
	r.mu.Unlock()

	if cached != nil && cached.done && cached.bestHost != nil {
		log.Write("Using cached CDN host for %s: %s", region, cached.bestHost.host)
		return cached.bestHost.host + serverConfig["Path"] + "/", nil
	}
	if cached != nil && cached.promise != nil {
		log.Write("Waiting for CDN resolution for %s", region)
		<-cached.promise
		r.mu.Lock()
		entry := r.resolutionCache[cacheKey]
		r.mu.Unlock()
		if entry.err != nil {
			return "", entry.err
		}
		return entry.rankedHosts[0].host + serverConfig["Path"] + "/", nil
	}

	log.Write("Resolving CDN hosts for %s: %s", region, serverConfig["Hosts"])
	entry := &resolutionCacheEntry{promise: make(chan struct{})}
	r.mu.Lock()
	r.resolutionCache[cacheKey] = entry
	r.mu.Unlock()

	rankedHosts, err := r.resolveHosts(region, serverConfig)
	if err != nil {
		entry.err = err
		close(entry.promise)
		entry.done = true
		return "", err
	}
	entry.rankedHosts = rankedHosts
	entry.bestHost = &rankedHosts[0]
	entry.done = true
	close(entry.promise)
	return rankedHosts[0].host + serverConfig["Path"] + "/", nil
}

// GetRankedHosts returns all available hosts ranked by ping speed.
func (r *CDNResolver) GetRankedHosts(region string, serverConfig VersionConfigEntry) ([]string, error) {
	cacheKey := r.cacheKey(region, serverConfig["Hosts"])
	r.mu.Lock()
	cached := r.resolutionCache[cacheKey]
	r.mu.Unlock()

	if cached != nil && cached.done && cached.rankedHosts != nil {
		log.Write("Using cached ranked CDN hosts for %s", region)
		return mapHosts(cached.rankedHosts, serverConfig["Path"]), nil
	}
	if cached != nil && cached.promise != nil {
		log.Write("Waiting for CDN resolution for %s", region)
		<-cached.promise
		r.mu.Lock()
		entry := r.resolutionCache[cacheKey]
		r.mu.Unlock()
		if entry.err != nil {
			return nil, entry.err
		}
		return mapHosts(entry.rankedHosts, serverConfig["Path"]), nil
	}

	log.Write("Resolving CDN hosts for %s: %s", region, serverConfig["Hosts"])
	entry := &resolutionCacheEntry{promise: make(chan struct{})}
	r.mu.Lock()
	r.resolutionCache[cacheKey] = entry
	r.mu.Unlock()

	rankedHosts, err := r.resolveHosts(region, serverConfig)
	if err != nil {
		entry.err = err
		close(entry.promise)
		entry.done = true
		return nil, err
	}
	entry.rankedHosts = rankedHosts
	entry.bestHost = &rankedHosts[0]
	entry.done = true
	close(entry.promise)
	return mapHosts(rankedHosts, serverConfig["Path"]), nil
}

// MarkHostFailed marks a host as failed.
func (r *CDNResolver) MarkHostFailed(host string) {
	log.Write("Marking CDN host as failed: %s", host)
	r.mu.Lock()
	r.failedHosts[host] = struct{}{}
	r.mu.Unlock()
}

// ClearCache clears resolver caches.
func (r *CDNResolver) ClearCache() {
	r.mu.Lock()
	r.resolutionCache = map[string]*resolutionCacheEntry{}
	r.failedHosts = map[string]struct{}{}
	r.mu.Unlock()
}

func (r *CDNResolver) cacheKey(region, hosts string) string {
	return region + "|" + hosts
}

func (r *CDNResolver) resolveHosts(region string, serverConfig VersionConfigEntry) ([]rankedHost, error) {
	log.Write("Resolving best host for %s: %s", region, serverConfig["Hosts"])
	hosts := strings.Fields(serverConfig["Hosts"])
	var validHosts []rankedHost
	var wg sync.WaitGroup
	var mu sync.Mutex

	for _, h := range hosts {
		host := "https://" + h + "/"
		r.mu.Lock()
		_, failed := r.failedHosts[host]
		r.mu.Unlock()
		if failed {
			log.Write("Skipping previously failed host: %s", host)
			continue
		}
		wg.Add(1)
		go func(h string) {
			defer wg.Done()
			pingMs, err := formats.Ping(h)
			if err != nil {
				log.Write("Host %s failed to resolve a ping: %s", h, err.Error())
				return
			}
			log.Write("Host %s resolved with %dms ping", h, pingMs)
			mu.Lock()
			validHosts = append(validHosts, rankedHost{host: h, ping: pingMs})
			mu.Unlock()
		}(host)
	}
	wg.Wait()
	if len(validHosts) == 0 {
		return nil, fmt.Errorf("unable to resolve any CDN hosts (all failed or blocked)")
	}
	sort.Slice(validHosts, func(i, j int) bool { return validHosts[i].ping < validHosts[j].ping })
	log.Write("%s resolved as the fastest host with a ping of %dms", validHosts[0].host, validHosts[0].ping)
	return validHosts, nil
}

func mapHosts(hosts []rankedHost, path string) []string {
	out := make([]string, len(hosts))
	for i, h := range hosts {
		out[i] = h.host + path + "/"
	}
	return out
}
